projects/igniteui-angular/src/lib/query-builder/query-builder-tree.component.ts

```diff
@@ -1412,9 +1425,73 @@ export class IgxQueryBuilderTreeComponent implements AfterViewInit, OnDestroy {
      * @hidden @internal
      */
     public enterExpressionEdit(expressionItem: ExpressionOperandItem) {
-        this.exitOperandEdit();
-        this.cancelOperandAdd();
+        if (this._editedExpression) {
+            this.exitOperandEdit();
+            this.cancelOperandAdd();
 
+            this._preventInit = true;
+        }
+        this.cdr.detectChanges();
+        this.enterEditMode(expressionItem);
+    }
+
+
+    /**
+     * @hidden @internal
+     */
+    public clickExpressionAdd(expressionItem: ExpressionOperandItem, targetButton: HTMLElement) {
+        if (this._editedExpression) {
+            this.exitOperandEdit();
+            this.cancelOperandAdd();
+
+            this._preventInit = true;
+        }
+        this.cdr.detectChanges();
+        this.openExpressionAddDialog(expressionItem, targetButton);
+    }
+
+    /**
+     * @hidden @internal
+     */
+    public openExpressionAddDialog(expressionItem: ExpressionOperandItem, targetButton: HTMLElement) {
```

Comment:
expressionItem is not used and can be removed from this method and from clickExpressionAdd(...)